### DISCUSSION TITLE
KXML2 NoClassDefFoundError

### DISCUSSION THREAD

**Discussion by tobithe1st** [mapsforge/mapsforge#1417] [Q&A] [answered]
I am getting this Exception although the "missing" classes are located in the dependency-jar-file... I already tried using different contributors of the kxml2 (also the newest and most maintained ones), but it didn't change anything...
```
[WARNING] There are 1 pathException(s). The related dependencies will be ignored.
[WARNING] Dependency: C:\Users\tobia\.m2\repository\net\sf\kxml\kxml2\2.3.0\kxml2-2.3.0.jar
   - exception: Unable to derive module descriptor for C:\Users\tobia\.m2\repository\net\sf\kxml\kxml2\2.3.0\kxml2-2.3.0.jar
   - cause: Provider class org.kxml2.io.KXmlParser,org.kxml2.io.KXmlSerializer not in module
Exception in thread "main" java.lang.NoClassDefFoundError: org/xmlpull/v1/XmlPullParserException
	at org.mapsforge.map.rendertheme.rule.RenderThemeFuture.<init>(RenderThemeFuture.java:70)
	at org.mapsforge.map.layer.renderer.TileRendererLayer.compileRenderTheme(TileRendererLayer.java:161)
	at org.mapsforge.map.layer.renderer.TileRendererLayer.setXmlRenderTheme(TileRendererLayer.java:157)
	at com.henrytriage.windows.scenarioMapWindow.Samples.addLayers(Samples.java:164)
	at com.henrytriage.windows.scenarioMapWindow.Samples.main(Samples.java:98)
Caused by: java.lang.ClassNotFoundException: org.xmlpull.v1.XmlPullParserException
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 5 more
```

**Comment by devemux86** [comment]
Android? Desktop?

The dependency is on Maven remote repository (not the local repository).
See also the [integration guide](https://github.com/mapsforge/mapsforge/blob/master/docs/Integration.md).

**Reply by tobithe1st** [reply]
Desktop.
I did everything mentioned in the integration guide. Also the kxml2.jar shows up in my dependencies and the class files are also located there...

**Reply by devemux86** [reply]
But your exception shows that it searches in the local Maven repository (not the remote one)?

What Java do you use? Could be some issue with Java or configuration of the dependencies.

**Reply by tobithe1st** [reply]
I now fixed this with adding the xmlpull 1.1.3.1 dependency.
But now I get the IllegalArgument exception every 0.1 second again, but more specific:
```
Juni 29, 2023 8:39:42 PM org.mapsforge.map.layer.renderer.DatabaseRenderer executeJob
WARNING: Exception: java.lang.IllegalArgumentException: Parse error for XML rendertheme
```
Application is running, but not displaying the .map file....

**Reply by devemux86** [reply]
The correct dependencies are mentioned in [integration guide](https://github.com/mapsforge/mapsforge/blob/master/docs/Integration.md).

The project is build successfully in [GitHub Actions](https://github.com/mapsforge/mapsforge/actions) and the samples work on Android and Desktop.

**Reply by tobithe1st** [reply]
Yes, I now have KXML2:2.3.0 AND xmlpull:1.1.3.1 and its working fine using the sample desktop-app, but only when setting SHOW_RASTER_MAP = true therefore using the online-map, but when setting it to false and passing in "args" the path to the local .map file, the IllegalArgumentException occurs for the parser.....

**Reply by devemux86** [reply]
We build and test on Java 11 with compiler 1.7 for maximum compatibility on Android and all users.

Why use different dependencies than those in the guide?
Is it a Java 9+ modules problem? What Java do you use?

If you are using a different Java or compiler, you will need to debug it, as the mentioned versions work without issues.

**Comment by devemux86** [comment, marked as answer]
I suggest to clone the Mapsforge repository and run the desktop sample.

I created the mapsforge-samples-awt-master-SNAPSHOT-jar-with-dependencies.jar, it contains all desktop dependencies (unzip it first):
[mapsforge-samples-awt-master-SNAPSHOT-jar-with-dependencies.jar.zip](https://github.com/mapsforge/mapsforge/files/11909797/mapsforge-samples-awt-master-SNAPSHOT-jar-with-dependencies.jar.zip)

See the [desktop guide](https://github.com/mapsforge/mapsforge/blob/master/docs/Desktop.md), you can run:
```
java -jar mapsforge-samples-awt-master-SNAPSHOT-jar-with-dependencies.jar andorra.map
```

Testing locally on Java 11, it loads the andorra.map and runs fine.
